rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,392 @@
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     ContainsDuplicates: true
+	}, nil
+}
+```
+
+### Removing an endpoint
+An endpoint can be removed after our backward compatibility promise has been
+fulfilled and NOT before. This means marking for deletion with a TODO and
+removing it in the relevant version. For example, if v17 no longer uses
+`GET /webapi/foos` , you can mark for deletion in v18, as v18 only needs to
+support n-1 versioned clients. It is the responsibility of the developer who is
+"removing the use" (i.e., making it obsolete with some other endpoint) to mark
+the TODO.
+
+
+### Adding/removing a field to a response of an existing endpoint
+Adding a new field to a response is OK as long as that field has to effect on
+the previously existing fields. Any field that was previously in a response
+should _stay populated_ in the new response, even if that creates duplicate
+data. An existing field _cannot_ have its type changed. A field should not be
+removed until the backward compatibility promise has been 'fulfilled'. If v17
+clients are only using a new field, you can mark the old field as "TODO DELETE"
+in v18. This means that v17 servers can support v16 clients (still using the old
+field) and v18 only needs to support v17 (using the new field).
+
+Bad
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    // ContainsDuplicates bool <--- previously existing field made redundant by DuplicateCount
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we don't need ContainsDuplcates anymore because now we can
+	     // just check if duplicate count > 0!"
+	     DuplicateCount: 1 
+	}, nil
+}
+```
+
+Good
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    // TODO (avatus) DELETE IN 18
+    ContainsDuplicates bool
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we must keep ContainsDuplicates populated to support older clients!"
+	     ContainsDuplicates: true
+	     DuplicateCount: 1 
+	}, nil
+}
+``` 
+If using ConnectRPC (a proposed solution later in this RFD), then we must _not_ remove fields from requests/responses, even if unused. Mark as deprecated and move on.
+
+If a response shape must change entirely, prefer creating a new endpoint/RPC.
+
+### Updating Request body fields
+A Request body can receive a new field as long as the omission of the new field
+would not cause a break (i.e., an old client sending the old request without the
+new field). If the handler _requires_ this new field in order to function, then
+a new endpoint must be created with a version increment.
```

Comment:
The request body could be something like this
```go
type MyRequest struct {
    Limit int
    Query string
}
```

And if we wanted to add a new field like
```go
type MyRequest struct {
    Limit int
    Query string
    MyNewThing string
}
```

if the handler has something along the lines of
```go
if req.MyNewThing == "" {
     return nil, err("BAD REQUEST")
}
```
that would be considered a breaking change because old clients do NOT send `MyNewThing` with their requests. therefore, a new endpoint is needed.

But, similar to the paginated access requests, we didn't need to change the endpoint because we added a new value `Limit` to the request, and if it wasn't present, we just considered it not paginated. This _wouldn't_ be a breaking change and should be allowed.

Reply:
Added this explaination snippet to the RFD